core/src/main/java/com/microsoft/applicationinsights/internal/agent/CoreAgentNotificationsHandler.java

```diff
@@ -0,0 +1,234 @@
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.internal.agent;
+
+import java.net.URL;
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.LinkedList;
+
+import com.microsoft.applicationinsights.agent.internal.coresync.AgentNotificationsHandler;
+import com.microsoft.applicationinsights.agent.internal.coresync.InstrumentedClassType;
+
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.internal.schemav2.DependencyKind;
+import com.microsoft.applicationinsights.telemetry.Duration;
+import com.microsoft.applicationinsights.telemetry.RemoteDependencyTelemetry;
+
+/**
+ * The Core's implementation: the methods are called for instrumented methods.
+ * The implementation can measure time in nano seconds, fetch Sql/Http data and report exceptions
+ *
+ * Created by gupele on 5/7/2015.
+ */
+final class CoreAgentNotificationsHandler implements AgentNotificationsHandler {
+
+    /**
+     * The class holds the data gathered on a method
+     */
+    private static class MethodData {
+        public String name;
+        public String[] arguments;
+        public long interval;
+        public InstrumentedClassType type;
+        public Object result;
+    }
+
+    private static class ThreadData {
+        public final LinkedList<MethodData> methods = new LinkedList<MethodData>();
+
+        public void done() {
+            methods.clear();
+        }
+    }
+
+    static final class ThreadDataThreadLocal extends ThreadLocal<ThreadData> {
+        private ThreadData threadData;
+
+        @Override
+        protected ThreadData initialValue() {
+            threadData = new ThreadData();
+            return threadData;
+        }
+    };
+
+    private ThreadDataThreadLocal threadDataThreadLocal = new ThreadDataThreadLocal();
+
+    private TelemetryClient telemetryClient = new TelemetryClient();
+
+    private final String name;
+
+    public CoreAgentNotificationsHandler(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public void onException(String className, String methodName, Throwable throwable) {
+        try {
+            if (throwable instanceof Exception) {
+                telemetryClient.trackException((Exception)throwable);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodEnterURL(String name, URL url) {
+        String urlAsString = url == null ? null : url.toString();
+        startMethod(InstrumentedClassType.HTTP, name, urlAsString);
+    }
+
+    @Override
+    public void onMethodEnterSqlStatement(String name, Statement statement, String sqlStatement) {
```

Comment:
This handler handle several types of data.
I think it would be more elegant delegating the handling of specific data types to specific handlers.
WDYT?